File: prof/src/com/oracle/souffleprof/Iteration.java
/*
 * Souffle - A Datalog Compiler
 * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved
 * Licensed under the Universal Permissive License v 1.0 as shown at:
 * - https://opensource.org/licenses/UPL
 * - <souffle root>/licenses/SOUFFLE-UPL.txt
 */

package com.oracle.souffleprof;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

/***
 * Profile Data Model
 * 
 * Represents recursive profile data 
 */

public class Iteration implements Serializable {

	private static final long serialVersionUID = 4513587123648922999L;
	private double runtime = 0;
	private long num_tuples = 0;
	private double copy_time = 0;
	private String locator = "";
	private long prev_num_tuples = 0;

	private Map<String, RuleRecursive> rul_rec_map;

	public Iteration() {
		this.rul_rec_map = new HashMap<String, RuleRecursive>();
	}

	public void addRule(String[] data, String rec_id) {
		String strTemp = data[4] + data[3] + data[2];

		if (data[0].charAt(0) == 't') {
			boolean exists = false;

			for (String name : rul_rec_map.keySet()) {
				if (name.equals(strTemp)) {
					RuleRecursive rul_rec = rul_rec_map.get(strTemp);
					rul_rec.setRuntime(Double.parseDouble(data[5])
							+ rul_rec.getRuntime());
					exists = true;
				}
			}
			if (!exists) {
				RuleRecursive rul_rec = new RuleRecursive(data[4],
						Integer.parseInt(data[2]), rec_id);
				rul_rec.setRuntime(Double.parseDouble(data[5]));
				rul_rec.setLocator(data[3]);
				rul_rec_map.put(strTemp, rul_rec);
			}

		} else if (data[0].charAt(0) == 'n') {
			RuleRecursive rul_rec = rul_rec_map.get(strTemp);
			assert rul_rec != null : "missing t tag";
			rul_rec.setNum_tuples(Long.parseLong(data[5]) - prev_num_tuples);
			this.prev_num_tuples = Long.parseLong(data[5]);
			rul_rec_map.put(strTemp, rul_rec);
		}
	}

	public Map<String, RuleRecursive> getRul_rec() {
		return rul_rec_map;
	}

	@Override
	public String toString() {

		StringBuilder res = new StringBuilder();
		res.append("" + runtime + "," + num_tuples + "," + copy_time + ",");
		res.append(" recRule:");
		for (RuleRecursive rul : rul_rec_map.values()) {
			res.append(rul.toString());
		}
		res.append("\n");
		return res.toString();
	}

	public double getRuntime() {
		return runtime;
	}

	public void setRuntime(double runtime) {
		this.runtime = runtime;
	}

	public long getNum_tuples() {
		return num_tuples;
	}

	public void setNum_tuples(long num_tuples) {
		this.num_tuples = num_tuples;
	}

	public double getCopy_time() {
		return copy_time;
	}

	public void setCopy_time(double copy_time) {
		this.copy_time = copy_time;
	}

	public String getLocator() {
		return locator;
	}

	public void setLocator(String locator) {
		this.locator = locator;
	}
}

File: src/profilerlib/Iteration.cpp
/*
* Souffle - A Datalog Compiler
* Copyright (c) 2016, The Souffle Developers. All rights reserved
* Licensed under the Universal Permissive License v 1.0 as shown at:
* - https://opensource.org/licenses/UPL
* - <souffle root>/licenses/SOUFFLE-UPL.txt
*/


#include <iostream>
#include "Iteration.hpp"


void Iteration::addRule(std::vector <std::string> data, std::string rec_id) {
    std::string strTemp = data[4] + data[3] + data[2];

    if (data[0].at(0) == 't') {

        if (rul_rec_map.find(strTemp) != rul_rec_map.end()) {
            std::shared_ptr <Rule> rul_rec = rul_rec_map[strTemp];
            rul_rec->setRuntime(std::stod(data[5]) + rul_rec->getRuntime());
        } else {
            std::shared_ptr <Rule> rul_rec = std::make_shared<Rule>(Rule(data[4],
                                                                         std::stoi(data[2]), rec_id));
            std::cout << "Iteration Rule runtime: " << (data[5]);
            std::cout << "\n\tstod(runtime): " << std::stod(data[5]) << std::endl;
            rul_rec->setRuntime(std::stod(data[5]));
            rul_rec->setLocator(data[3]);
            rul_rec_map[strTemp] = rul_rec;
        }

    } else if (data[0].at(0) == 'n') {
        std::unordered_map < std::string, std::shared_ptr < Rule >> ::const_iterator
        got = rul_rec_map.find(strTemp);

        assert(got != rul_rec_map.end() && "missing t tag");
        std::shared_ptr <Rule> rul_rec = rul_rec_map[strTemp];
        rul_rec->setNum_tuples(std::stol(data[5]) - prev_num_tuples);
        this->prev_num_tuples = std::stol(data[5]);
        rul_rec_map[strTemp] = rul_rec;

    }
}

std::string Iteration::toString() {
    std::ostringstream output;

    output << runtime << "," << num_tuples << "," << copy_time << ",";
    output << " recRule:";
    for (auto &rul : rul_rec_map)
        output << rul.second->toString();
    output << "\n";
    return output.str();
}